sql/src/main/java/io/crate/planner/operators/Order.java

```diff
@@ -101,9 +102,45 @@ public ExecutionPlan build(PlannerContext plannerContext,
         return executionPlan;
     }
 
+    @Override
+    public LogicalPlan tryOptimize(@Nullable LogicalPlan pushDown) {
+        if (pushDown instanceof Order) {
+            // We can combine the order being pushed with this Order
+            OrderBy mergedOrderBy = ((Order) pushDown).orderBy.merge(orderBy);
```

Comment:
Do we have to merge it? Isn't it safe to replace it entirely?

```
Order [y]
  |
Order [x]
```
Should be safe to only order by x, though I'm not sure about cases like:

```
Order by [x]
  |
Order by [x, y]
```

Reply:
Postgres seems to do 2 sorts:
```explain select * from (select * from t1 order by a, b) a
union all
select * from t1
order by 1```
=>
```'Merge Append  (cost=317.02..384.82 rows=4520 width=8)'
'  Sort Key: t1_1.a'
'  ->  Sort  (cost=158.51..164.16 rows=2260 width=8)'
'        Sort Key: t1_1.a, t1_1.b'
'        ->  Seq Scan on t1 t1_1  (cost=0.00..32.60 rows=2260 width=8)'
'  ->  Sort  (cost=158.51..164.16 rows=2260 width=8)'
'        Sort Key: t1.a'
'        ->  Seq Scan on t1  (cost=0.00..32.60 rows=2260 width=8)'```

So I guess we need to merge the ORDER BYs so that we keep the semantics.

Reply:
That was just my initial attempt but you're right that we can further optimize this.

>Should be safe to only order by x, though I'm not sure about cases like:
```
Order by [x]
  |
Order by [x, y]
```

What cases are you thinking about? For Union that should always work.

Reply:
In this second case if we simply dropped the child ordering we'd miss that it is pre-sorted by y as well.
Though I'm not sure if this is a required semantic.

Reply:
My opinion is to keep the merge which keeps the semantics: https://github.com/crate/crate/blob/mxm/union-pushdown2/sql/src/test/java/io/crate/analyze/OrderByTest.java#L75

Reply:
We decide to replace the order by.